lib/Canonicalize/SimplifySwitchPass.cpp

```diff
@@ -0,0 +1,263 @@
+                                     llvm::Constant *C) {
+  for (const auto &I : Switch->cases()) {
+    const llvm::ConstantInt *CaseValue = I.getCaseValue();
+    if (CaseValue == C)
+      return I.getCaseSuccessor();
+  }
+
+  return nullptr;
+}
+
+static bool runSimplifySwitch(llvm::Function &F,
+                              llvm::LazyValueInfo &LVI,
+                              llvm::DominatorTree &DT,
+                              RawBinaryMemoryOracle &MO) {
+  bool Result = false;
+  llvm::SmallVector<llvm::SwitchInst *, 16> SwitchToDelete;
+  for (llvm::BasicBlock &BB : F) {
+    for (llvm::Instruction &I : BB) {
+      llvm::SwitchInst *Switch = dyn_cast<llvm::SwitchInst>(&I);
+      if (not Switch)
+        continue;
+
+      // Skip empty switches.
+      if (Switch->getNumCases() == 0)
+        continue;
+
+      auto *Condition = Switch->getCondition();
+      DataFlowGraph::Limits Limits(1000 /*MaxPhiLike*/, 1 /*MaxLoad*/);
+      auto Results = ValueMaterializer::getValuesFor(Switch,
+                                                     Condition,
+                                                     MO,
+                                                     LVI,
+                                                     DT,
+                                                     Limits,
+                                                     Oracle::AdvancedValueInfo);
+      // We can not do too much without any values materialized.
+      if (not Results.values())
+        continue;
+
+      // The number of values materialized should match the number of cases in
+      // the switch.
+      if (Results.values()->size() != Switch->getNumCases()) {
+        revng_log(Log, "Unexpected number of values found!");
+        continue;
+      }
+
+      auto &DFG = Results.dataFlowGraph();
+      // Identify the only node in the data flow graph that's associated to an
+      // Oracle and use it as a switch condition.
+      bool BailOut = false;
+      BidirectionalNode<DataFlowNode> *StartNode = nullptr;
+
+      for (auto BidirectNode : DFG.nodes()) {
+        const DataFlowNode &Node = *BidirectNode;
+        // Ignore nodes that are constants in the DFG.
+        if (llvm::isa<llvm::ConstantInt>(Node.Value))
+          continue;
+
+        if (Node.UseOracle) {
+          if (StartNode != nullptr) {
+            BailOut = true;
+            break;
+          } else {
+            // TODO: Fix this const_cast.
+            BidirectionalNode<DataFlowNode>
+              *N = const_cast<BidirectionalNode<DataFlowNode> *>(BidirectNode);
+            StartNode = N;
+          }
+        }
+      }
+
+      // We found more than one start node.
+      if (StartNode == nullptr or BailOut)
+        continue;
+
+      std::map<llvm::ConstantInt *, llvm::BasicBlock *> NewLabels;
+      for (auto &Value : *StartNode->OracleRange) {
+        auto Old = DFG.materializeOne(StartNode, MO, Value);
+        if (not Old)
+          continue;
```

Comment:
1. This and the next `continue` work, but if you outline the body of the `for` into a function you can turn these into a `return`, which is better.
2. For each failing case, print a diagnosis to the logger.

Reply:
Not fixed